src/index.js

```diff
@@ -1,47 +1,118 @@
-/** @license MIT License (c) copyright 2015-2017 original author or authors */
-
+// @flow
+import type { Stream, Sink, ScheduledTask, Time, Scheduler, Disposable } from '@most/types'
 import { MulticastSource } from '@most/core'
+import { asap, cancelTask } from '@most/scheduler'
+
+export const hold = <A> (stream: Stream<A>): Stream<A> =>
+  new Hold(stream)
+
+type HeldValue<A> = { value: A }
 
-// hold :: Stream a -> Stream a
-export const hold = stream =>
-  new MulticastSource(new Hold(stream))
+class Hold<A> extends MulticastSource<A> {
+  pendingSinks: Sink<A>[]
+  held: ?HeldValue<A>
+  task: ?ScheduledTask
+
+  constructor (source: Stream<A>) {
+    super(source)
+    this.pendingSinks = []
+    this.held = undefined
+    this.task = undefined
+  }
+
+  run (sink: Sink<A>, scheduler: Scheduler): Disposable {
+    if (this._shouldScheduleFlush()) {
+      this._scheduleFlush(sink, scheduler)
+    }
+
+    return super.run(sink, scheduler)
+  }
+
+  _hasValue (): boolean {
+    return this.held !== undefined
+  }
+
+  _hasSinks (): boolean {
+    return this.sinks.length > 0
+  }
+
+  _shouldScheduleFlush (): boolean {
+    return this._hasValue() && this._hasSinks()
+  }
+
+  dispose (): void {
+    this._cancelTask()
+    return super.dispose()
+  }
 
-class Hold {
-  constructor (source) {
-    this.source = source
-    this.time = -Infinity
-    this.value = void 0
+  event (time: Time, value: A): void {
+    const pendingSinks = this._flushPending(time)
+    this.sinks = this.sinks.concat(pendingSinks)
+    this.held = { value }
```

Comment:
This is unfortunate, and I'd love to find a way to avoid an object allocation per event (without also introducing more mutation!).  The loop on line 60 below, however, needs to know *for sure* that there is a held value (remembering that `null` and `undefined` are perfectly legal event values).  AFAIK, there's no way, in a single class, with flow to have a boolean field that represents whether another field has a value--IOW, I don't think there's a way to represent that relationship in a way the type checker understands.

Def open to ideas!

Reply:
I only see this possible with mutation. I also don’t have enough knowledge about flow.

Reply:
> I only see this possible with mutation

Hmm, I didn't quite understand.  Do you have a suggestion for improving this, or are you also lamenting the need for the `{ value }` object as I am?

I'd be interested to see if there a solution to this problem in TypeScript.  Anyone know?

Reply:
I’d have to play around a bit.

Reply:
What if we introduced a private reference boolean that is set `event`? Perhaps `_hasValue` could be a property rather than the current method that now checks `this.held`. Then check `this._hasValue` in `_flushPending` before using `tryEvent` with `this.held`. Then you don’t need an object. Or am I missing something?

Reply:
TypeScript has about the same level of support for this.

Reply:
Given our slack discussion, it seems we still don't know of a solution we like better.  I'm in favor of merging as is, with the intent to improve it later if/when we discover a better way.